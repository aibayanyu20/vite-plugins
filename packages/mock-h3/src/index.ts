import type { PluginOption } from 'vite'
import { createLogger } from 'vite'
import { mockServer } from './server'

export const logger = createLogger('info', {
  prefix: '[vite:mock]',
})

export interface ViteMockContext extends MockPluginOptions {
  registeredRoutes: Set<string>
}

export interface MockPluginOptions {
  /**
   * 定义插件的前缀
   * @default '/mock'
   */
  prefix?: string

  /**
   * 配置mock的目录
   * @default 'mock'
   */
  mockDir?: string
}

export function mockPlugin(options: MockPluginOptions = { prefix: '/mock', mockDir: 'mock' }): PluginOption {
  const ctx: ViteMockContext = {
    registeredRoutes: new Set<string>(),
    ...options,
  }

  return {
    name: 'vite:mock',
    async configureServer(server) {
      await mockServer(server, ctx)
    },
    async configurePreviewServer(server) {
      await mockServer(server, ctx)
    },
  }
}

export default mockPlugin
