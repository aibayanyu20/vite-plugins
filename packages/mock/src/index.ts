;(async () => {
  try {
    await import('mockjs')
  } catch (e) {
    throw new Error('vite-plugin-vue-mock requires mockjs to be present in the dependency tree.')
  }
})()

import type { ViteMockOptions } from './types'
import type { Plugin } from 'vite'
import { ResolvedConfig } from 'vite'
import { createMockServer, requestMiddleware } from './createMockServer'

export function viteMockServe(opt: ViteMockOptions = {}): Plugin {
  let isDev = false
  let config: ResolvedConfig

  return {
    name: 'vite:mock',
    enforce: 'pre' as const,
    configResolved(resolvedConfig) {
      config = resolvedConfig
      isDev = config.command === 'serve'
      isDev && createMockServer(opt, config)
    },

    configureServer: async ({ middlewares }) => {
      const { enable = isDev } = opt
      if (!enable) {
        return
      }
      const middleware = await requestMiddleware(opt)
      middlewares.use(middleware)
    },
  }
}

export * from './types'
