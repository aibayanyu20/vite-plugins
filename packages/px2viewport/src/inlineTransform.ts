import { parse } from '@babel/parser'
import type { ObjectExpression } from '@babel/types'
import { traverse } from './utils/traverse'
import { generate } from './utils/genrate'

const ReU = /(\d+)px/g

function pxToVw(input: string, baseWidth: number = 750): string {
  return input.replace(ReU, (_match, pxValue) => {
    const vwValue = Number(((Number.parseInt(pxValue) / baseWidth) * 100).toFixed(5))
    return `${vwValue}vw`
  })
}
function resolveStyleObject(node: ObjectExpression) {
  const props = node.properties
  for (const prop of props) {
    if (prop.type === 'ObjectProperty' && prop.value.type === 'StringLiteral') {
      if (prop.value.value)
        prop.value.value = pxToVw(prop.value.value)
    }
  }
}

function resolveObjectExpression(node: ObjectExpression) {
  const props = node.properties
  for (const prop of props) {
    if (prop.type === 'ObjectProperty' && prop.key.type === 'Identifier' && prop.key.name === 'style') {
      if (prop.value.type === 'ObjectExpression')
        resolveStyleObject(prop.value)
    }
  }
}

export function inlineTransform(code: string) {
  // 使用babel进行转换
  const ast = parse(code, {
    sourceType: 'module',
  })
  traverse(ast, {
    // 获取当前定义的函数
    VariableDeclarator({ node }) {
      if (node.id) {
        // 判断一下name是不是为_hoisted_变量提升的数据
        if (node.id.type === 'Identifier' && node.id.name.startsWith('_hoisted_')) {
          // 证明是变量提升的数据
          if (node.init && node.init.type === 'ObjectExpression') {
            // 如果是一个对象的情况下如何进行处理
            resolveObjectExpression(node.init)
          }
        }
      }
    },
  })
  return generate(ast).code
}
