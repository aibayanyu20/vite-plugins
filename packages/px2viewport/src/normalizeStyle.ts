type NormalizedStyle = Record<string, string | number>
const isArray = Array.isArray
const isString = (val: unknown): val is string => typeof val === 'string'
const isObject = (val: unknown): val is Record<string, unknown> => val !== null && typeof val === 'object'

const listDelimiterRE = /;(?![^(]*\))/g
const propertyDelimiterRE = /:([^]+)/
const styleCommentRE = /\/\*[^]*?\*\//g

function parseStringStyle(cssText: string): NormalizedStyle {
  const ret: NormalizedStyle = {}
  cssText
    .replace(styleCommentRE, '')
    .split(listDelimiterRE)
    .forEach((item) => {
      if (item) {
        const tmp = item.split(propertyDelimiterRE)
        tmp.length > 1 && (ret[tmp[0].trim()] = tmp[1].trim())
      }
    })
  return ret
}

interface Px2viewportOptions {
/**
 * 需要转换的单位，默认为"px"
 */
  unitToConvert?: string
  /**
   * 单位转换后保留的精度
   */
  unitPrecision?: number
  /**
   * 希望使用的视口单位
   */
  viewportUnit?: string
  /**
   * 设置最小的转换数值，如果为1的话，只有大于1的值会被转换
   */
  minPixelValue?: number
  /**
   * 设计稿的视口宽度
   */
  viewportWidth?: number
}
declare global {
  interface Window {
    __px2viewport: Px2viewportOptions
  }
}

function pxToVw(input: string): string {
  let config: Px2viewportOptions = {
    viewportWidth: 750,
    unitToConvert: 'px',
    unitPrecision: 5,
    viewportUnit: 'vw',
    minPixelValue: undefined,
  }
  if (window && window.__px2viewport) {
    config = {
      ...config,
      ...window.__px2viewport,
    }
  }
  const baseWidth = config.viewportWidth as number
  const ReU = new RegExp(`(\\d+)${config.unitToConvert}`, 'g')
  return input.replace(ReU, (_match, pxValue) => {
    const vwValue = Number(((Number.parseInt(pxValue) / baseWidth) * 100).toFixed(config.unitPrecision ?? 5))
    if (config.minPixelValue && vwValue < config.minPixelValue)
      return `${config.minPixelValue}${config.unitToConvert}`

    return `${vwValue}${config.viewportUnit}`
  })
}

export function formatStyleValue(value: any) {
  // 检查末位是不是px，如果是px的情况下，就对数据进行处理
  if (isString(value))
    return pxToVw(value)

  return value
}

export function normalizeStyle(value: unknown): NormalizedStyle | string | undefined {
  if (isArray(value)) {
    const res: NormalizedStyle = {}
    for (let i = 0; i < value.length; i++) {
      const item = value[i]
      const normalized = isString(item)
        ? parseStringStyle(item)
        : (normalizeStyle(item) as NormalizedStyle)
      if (normalized) {
        for (const key in normalized)
          res[key] = formatStyleValue(normalized[key])
      }
    }
    return res
  }
  else if (isString(value)) {
    const res: NormalizedStyle = {}
    const normalized = parseStringStyle(value) ?? {}
    for (const key in normalized)
      res[key] = formatStyleValue(normalized[key])
    return res
  }
  else if (isObject(value)) {
    const res: NormalizedStyle = {}
    for (const key in value) {
      const val = formatStyleValue(value[key])
      res[key] = val as any
    }
    return res
  }
}
