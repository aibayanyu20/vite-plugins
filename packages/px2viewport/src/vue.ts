export { normalizeStyle, mergeProps } from './normalizeStyle'
