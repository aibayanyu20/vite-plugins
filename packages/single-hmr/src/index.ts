import type { Plugin } from 'vite'

export const VitePluginSingleHmr = (): Plugin => {
  return {
    name: 'vite:single-hmr',
    handleHotUpdate({ modules }) {
      modules.forEach((m) => {
        m.importedModules = new Set()
        m.importers = new Set()
      })
      return modules
    },
  }
}

export default VitePluginSingleHmr
