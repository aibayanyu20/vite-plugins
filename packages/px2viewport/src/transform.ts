import { findStaticImports } from 'mlly'
import { MagicString } from 'magic-string-ast'
import { inlineTransform } from './inlineTransform'
import type { Px2viewportOptionsCommon } from './index'

export function transform(code: string, config?: Px2viewportOptionsCommon): undefined | any {
  code = inlineTransform(code, config)
  const m = new MagicString(code)
  // 处理全局的参数来实现px2viewport
  const imports = findStaticImports(code)
  for (const source of imports) {
    if (source.specifier !== '@mistjs/vite-plugin-px2viewport/vue') {
      // 判断里面是否包含 normalizeStyle as _normalizeStyle,
      // 如果包含就去掉，然后用自己定义的 normalizeStyle
      let _imports = source.imports
      const normalizeStyleStr = 'normalizeStyle as _normalizeStyle'
      const normalizeStyleStr2 = 'normalizeStyle'
      const mergePropsStr = 'mergeProps as _mergeProps'
      const mergePropsStr2 = 'mergeProps'
      const px2viewportImport = []
      if (_imports && _imports.includes(normalizeStyleStr)) {
        let replaceCode = normalizeStyleStr
        if (_imports.includes(`${normalizeStyleStr},`)) {
          // 有逗号的情况
          replaceCode = `${normalizeStyleStr},`
        }
        else if (_imports.includes(`, ${normalizeStyleStr}`)) {
          // 有逗号的情况
          replaceCode = `, ${normalizeStyleStr}`
        }
        _imports = _imports.replace(replaceCode, '')
        px2viewportImport.push(normalizeStyleStr)
      }
      else if (_imports && _imports.includes(normalizeStyleStr2)) {
        let replaceCode = normalizeStyleStr2
        if (_imports.includes(`${normalizeStyleStr2},`)) {
          // 有逗号的情况
          replaceCode = `${normalizeStyleStr2},`
        }
        else if (_imports.includes(`, ${normalizeStyleStr2}`)) {
          // 有逗号的情况
          replaceCode = `, ${normalizeStyleStr2}`
        }
        _imports = _imports.replace(replaceCode, '')
        px2viewportImport.push(normalizeStyleStr2)
      }
      if (_imports && _imports.includes(mergePropsStr)) {
        let replaceCode = mergePropsStr
        if (_imports.includes(`${mergePropsStr},`)) {
          // 有逗号的情况
          replaceCode = `${mergePropsStr},`
        }
        else if (_imports.includes(`, ${mergePropsStr}`)) {
          // 有逗号的情况
          replaceCode = `, ${mergePropsStr}`
        }
        _imports = _imports.replace(replaceCode, '')
        px2viewportImport.push(mergePropsStr)
      }
      else if (_imports && _imports.includes(mergePropsStr2)) {
        let replaceCode = mergePropsStr2
        if (_imports.includes(`${mergePropsStr2},`)) {
          // 有逗号的情况
          replaceCode = `${mergePropsStr2},`
        }
        else if (_imports.includes(`, ${mergePropsStr2}`)) {
          // 有逗号的情况
          replaceCode = `, ${mergePropsStr2}`
        }
        _imports = _imports.replace(replaceCode, '')
        px2viewportImport.push(mergePropsStr2)
      }
      if (px2viewportImport.length) {
        const replaceImport = `import ${_imports} from '${source.specifier}'\nimport { ${px2viewportImport.join(', ')} } from '@mistjs/vite-plugin-px2viewport/vue';\n`
        m.overwrite(source.start, source.end, replaceImport)
      }
    }
  }
  return {
    code: m.toString(),
    map: m.generateMap({ hires: true }),
  }
}
