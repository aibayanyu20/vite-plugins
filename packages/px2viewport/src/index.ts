import type { FilterPattern, PluginOption } from 'vite'
import { createFilter } from 'vite'
import viewPort from 'postcss-px-to-viewport-8-plugin'
import { transform } from './transform'

export interface Px2viewportCSSOptions {
  /**
   * 设计稿的视口高度
   */
  viewportHeight?: number
  /**
   * 字体使用的视口单位
   */
  fontViewportUnit?: string
  /**
   * 需要忽略的CSS选择器，不会转为视口单位，使用原有的px等单位
   * 如果传入的值为字符串的话，只要选择器中含有传入值就会被匹配：例如 selectorBlackList 为 ['body'] 的话， 那么 .body-class 就会被忽略
   * 如果传入的值为正则表达式的话，那么就会依据CSS选择器是否匹配该正则：例如 selectorBlackList 为 [/^body$/] , 那么 body 会被忽略，而 .body 不会
   */
  selectorBlackList?: string[]
  /**
   * 能转化为vw的属性列表
   * 传入特定的CSS属性
   * 可以传入通配符""去匹配所有属性，例如：['']
   * 在属性的前或后添加"*",可以匹配特定的属性. (例如['position'] 会匹配 background-position-y)
   * 在特定属性前加 "!"，将不转换该属性的单位 . 例如: ['*', '!letter-spacing']，将不转换letter-spacing
   * "!" 和 ""可以组合使用， 例如: ['', '!font*']，将不转换font-size以及font-weight等属性
   */
  propList?: string[]

  /**
   * 媒体查询里的单位是否需要转换单位
   */
  mediaQuery?: boolean
  /**
   * 是否直接更换属性值，而不添加备用属性
   */
  replace?: boolean
  /**
   * 忽略某些文件夹下的文件或特定文件，例如 'node_modules' 下的文件
   * 如果值是一个正则表达式，那么匹配这个正则的文件会被忽略
   * 如果传入的值是一个数组，那么数组里的值必须为正则
   */
  exclude?: RegExp | RegExp[]
  /**
   * 如果设置了include，那将只有匹配到的文件才会被转换
   * 如果值是一个正则表达式，将包含匹配的文件，否则将排除该文件
   * 如果传入的值是一个数组，那么数组里的值必须为正则
   */
  include?: RegExp | RegExp[]
  /**
   * 是否添加根据 landscapeWidth 生成的媒体查询条件 @media (orientation: landscape)
   */
  landscape?: boolean
  /**
   * 横屏时使用的单位
   */
  landscapeUnit?: string
  /**
   * 横屏时使用的视口宽度
   * 支持传入函数，函数的参数为当前处理的文件路径
   */
  landscapeWidth?: number | ((filePath: string) => number | undefined)
}

export interface Px2viewportOptionsCommon {
  /**
   * 需要转换的单位，默认为"px"
   */
  unitToConvert?: string
  /**
   * 单位转换后保留的精度
   */
  unitPrecision?: number
  /**
   * 希望使用的视口单位
   */
  viewportUnit?: string
  /**
   * 设置最小的转换数值，如果为1的话，只有大于1的值会被转换
   */
  minPixelValue?: number
  /**
   * 设计稿的视口宽度
   */
  viewportWidth?: number | ((file: string) => number)
}

export interface Px2viewportOptions extends Px2viewportOptionsCommon {
  include?: FilterPattern
  cssOptions?: Px2viewportCSSOptions
}

export function px2viewport(options: Px2viewportOptions = {
  viewportWidth: 750,
  include: [/\.vue/, /\.[jt]sx$/],
}): PluginOption {
  const filter = createFilter(options?.include ?? [/\.vue/, /\.[jt]sx/])

  const commonConfig: Px2viewportOptionsCommon = {
    unitToConvert: options.unitToConvert ?? 'px',
    unitPrecision: options.unitPrecision ?? 5,
    viewportUnit: options.viewportUnit ?? 'vw',
    minPixelValue: options.minPixelValue,
    viewportWidth: options.viewportWidth ?? 750,
  }
  return {
    name: 'px2viewport',
    enforce: 'post',
    transform(code, id) {
      if (filter(id)) {
        let num = 750
        if (typeof options.viewportWidth === 'function')
          num = options.viewportWidth(id) ?? 750
        else
          num = options.viewportWidth ?? 750

        const res = transform(code, { ...commonConfig, viewportWidth: num })
        if (res)
          return res
      }
    },
    transformIndexHtml(html) {
      const _config = {
        ...commonConfig,
      }
      if (typeof commonConfig.viewportWidth === 'function')
        _config.viewportWidth = commonConfig.viewportWidth('px2viewport-inline')

      return {
        html,
        tags: [
          {
            tag: 'script',
            children: `window.__px2viewport = ${JSON.stringify(_config)}`,
          },
        ],
      }
    },
    config() {
      return {
        css: {
          postcss: {
            plugins: [
              viewPort({
                ...commonConfig,
                ...options?.cssOptions,
              }),
            ],
          },
        },
      }
    },
  } as PluginOption
}
