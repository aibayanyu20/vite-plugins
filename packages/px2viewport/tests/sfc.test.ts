import { describe, expect, it } from 'vitest'
import { inlineTransform } from '../src/inlineTransform'
import { sfcCode } from './fixtrures/sfc-code'
import { compCode } from './fixtrures/comp-code'
import { mergePropsCode, mergePropsCode1 } from './fixtrures/merge-props-code'

describe('sfc', () => {
  it('should work', () => {
    const code = inlineTransform(sfcCode)
    expect(code).toMatchSnapshot()
    const compCode1 = inlineTransform(compCode)
    expect(compCode1).toMatchSnapshot()
    const _mergePropsCode1 = inlineTransform(mergePropsCode)
    expect(_mergePropsCode1).toMatchSnapshot()
    const mergePropsCode2 = inlineTransform(mergePropsCode1)
    expect(mergePropsCode2).toMatchSnapshot()
  })
})
