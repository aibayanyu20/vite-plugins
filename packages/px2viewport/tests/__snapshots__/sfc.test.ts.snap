// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sfc > should work 1`] = `
"const __sfc__ = {};
import { createElementVNode as _createElementVNode, createTextVNode as _createTextVNode, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
function render(_ctx, _cache) {
  return _openBlock(), _createElementBlock("div", null, _cache[0] || (_cache[0] = [_createTextVNode(" 我的 "), _createElementVNode("div", {
    class: "bg-red",
    style: {
      "width": "13.33333vw",
      "height": "13.33333vw"
    }
  }, " Test ", -1 /* HOISTED */)]));
}

__sfc__.render = render;
__sfc__.__file = "src/App.vue";
export default __sfc__;"
`;

exports[`sfc > should work 2`] = `
"import { ref } from 'vue';
import Comp from './Comp.vue';
const __sfc__ = {
  __name: 'App',
  setup(__props, {
    expose: __expose
  }) {
    __expose();
    const msg = ref('Hello World!');
    const __returned__ = {
      msg,
      ref,
      Comp
    };
    Object.defineProperty(__returned__, '__isScriptSetup', {
      enumerable: false,
      value: true
    });
    return __returned__;
  }
};
import { toDisplayString as _toDisplayString, createElementVNode as _createElementVNode, vModelText as _vModelText, withDirectives as _withDirectives, createVNode as _createVNode, Fragment as _Fragment, openBlock as _openBlock, createElementBlock as _createElementBlock } from "vue";
function render(_ctx, _cache, $props, $setup, $data, $options) {
  return _openBlock(), _createElementBlock(_Fragment, null, [_createElementVNode("h1", null, _toDisplayString($setup.msg), 1 /* TEXT */), _withDirectives(_createElementVNode("input", {
    "onUpdate:modelValue": _cache[0] || (_cache[0] = $event => $setup.msg = $event)
  }, null, 512 /* NEED_PATCH */), [[_vModelText, $setup.msg]]), _createVNode($setup["Comp"], {
    style: {
      "height": "13.33333vw"
    }
  })], 64 /* STABLE_FRAGMENT */);
}

__sfc__.render = render;
__sfc__.__file = "src/App.vue";
export default __sfc__;"
`;

exports[`sfc > should work 3`] = `
"/* Analyzed bindings: {
  "attrss": "setup-const",
  "tableData": "setup-const"
} */
import { defineComponent as _defineComponent } from 'vue';
import { resolveComponent as _resolveComponent, createVNode as _createVNode, mergeProps as _mergeProps, withCtx as _withCtx, openBlock as _openBlock, createBlock as _createBlock } from "vue";
const __sfc__ = /*@__PURE__*/_defineComponent({
  __name: 'App',
  setup(__props) {
    const attrss = {};
    const tableData = [{
      date: '2016-05-03',
      name: 'Tom',
      address: 'No. 189, Grove St, Los Angeles'
    }, {
      date: '2016-05-02',
      name: 'Tom',
      address: 'No. 189, Grove St, Los Angeles'
    }, {
      date: '2016-05-04',
      name: 'Tom',
      address: 'No. 189, Grove St, Los Angeles'
    }, {
      date: '2016-05-01',
      name: 'Tom',
      address: 'No. 189, Grove St, Los Angeles'
    }];
    return (_ctx, _cache) => {
      const _component_el_table_column = _resolveComponent("el-table-column");
      const _component_el_table = _resolveComponent("el-table");
      return _openBlock(), _createBlock(_component_el_table, _mergeProps({
        data: tableData,
        style: {
          "width": "100px"
        }
      }, attrss), {
        default: _withCtx(() => [_createVNode(_component_el_table_column, {
          prop: "date",
          label: "Date",
          width: "180"
        }), _createVNode(_component_el_table_column, {
          prop: "name",
          label: "Name",
          width: "180"
        }), _createVNode(_component_el_table_column, {
          prop: "address",
          label: "Address"
        })]),
        _: 1 /* STABLE */
      }, 16 /* FULL_PROPS */);
    };
  }
});

__sfc__.__file = "src/App.vue";
export default __sfc__;"
`;

exports[`sfc > should work 4`] = `"import { Comment, Fragment, NOOP, Teleport, Text, Transition, TransitionGroup, __commonJS, __toESM, camelize, capitalize, cloneVNode, computed, createApp, createBaseVNode, createBlock, createCommentVNode, createElementBlock, createSlots, createTextVNode, createVNode, customRef, defineComponent, effectScope, getCurrentInstance, getCurrentScope, guardReactiveProps, h, hasOwn, hyphenate, inject, isArray, isDate, isFunction, isObject, isPlainObject, isPromise, isRef, isString, isVNode, markRaw, mergeProps, nextTick, normalizeClass, normalizeProps, normalizeStyle, onActivated, onBeforeMount, onBeforeUnmount, onBeforeUpdate, onDeactivated, onMounted, onScopeDispose, onUnmounted, onUpdated, openBlock, provide, reactive, readonly, ref, render, renderList, renderSlot, resolveComponent, resolveDirective, resolveDynamicComponent, shallowReactive, shallowRef, toDisplayString, toHandlerKey, toHandlers, toRaw, toRef, toRefs, unref, useAttrs, useSlots, vModelCheckbox, vModelRadio, vModelText, vShow, warn, watch, watchEffect, withCtx, withDirectives, withKeys, withModifiers } from "/node_modules/.vite/deps/chunk-ZHP7DJLA.js?v=2689b94e";"`;
